Create Topic Editor (Part 3): Create the editor UI

aks681 opened the pull request:
In this PR:

- TopicRights have been added, along with its corresponding services.
- A method to publish and unpublish topics have been added.
- A basic UI with functionality to edit Name and Description are done.
- Have a Save Topic functionality.

The editor can be accessed by clicking on the topic to edit in the Topics and Skills Dashboard.
The commits for this PR are from (and including) e3449b1.

Current UI:
![screenshot from 2018-06-16 20-51-10](https://user-images.githubusercontent.com/22347970/41499954-19595ddc-71a7-11e8-9b6f-918b8b54007d.png)

aks681 committed:
Merged with topics and skills dashboard

aks681 committed:
Created editor

aks681 committed:
disabled new structures

aks681 committed:
fixed merge conflicts

aks681 committed:
fixed errors

aks681 committed:
Added topic rights

aks681 committed:
fixed errors

aks681 committed:
minor change

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +227,37 @@ def delete(self, topic_id):
+    @acl_decorators.can_manage_rights_for_topic
+    def get(self, topic_id):
+        """Returns the TopicRights object of a topic.
+        """
```
Move to the end of the previous line.

/cc @aks681 @kevinlee12

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +227,37 @@ def delete(self, topic_id):
+class TopicRightsHandler(base.BaseHandler):
+    """A handler for returning topic rights."""
+
+    @acl_decorators.can_manage_rights_for_topic
```
I don't entirely understand the point of this handler. Is it for viewing rights or for managing rights? I don't see any managing going on here...

aks681 replied:
It's just for viewing rights, or, shall I merge TopicManagerRightsHandler and this? I didn't do this because the url for the former is /assign_manager. Shall I change the url to /manage_topic_rights and merge them?

aks681 replied:
Also, as all those who can edit the topic should be able to get the rights object, changed the rule to can_edit_topic.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +227,37 @@ def delete(self, topic_id):
+            if topic_services.check_can_edit_topic(
+                    user_actions_info, topic_rights):
+                can_edit_topic = True
+        except Exception as e:
```
This is a bit broad. Why would an exception occur?

aks681 replied:
Yeah, the only error is when topic rights doesn't exist, will change that.

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +227,37 @@ def delete(self, topic_id):
+            can_edit_topic = False
+            if topic_services.check_can_edit_topic(
+                    user_actions_info, topic_rights):
+                can_edit_topic = True
```
Should the previous four lines be replaced with:

    can_edit_topic = topic_services.check_can_edit_topic(...)

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +227,37 @@ def delete(self, topic_id):
+            'is_published': topic_rights.topic_is_published
+        }
+        self.values.update({
+            'topic_rights': topic_rights_to_return
```
I'm not sure about the need for the extra indirection. Perhaps just return the dict with topic_id, can_edit_topic, etc. instead?

Also, not sure why we need to send the topic_id back. Presumably the frontend already knows it.

aks681 replied:
Ok, yeah, removed topic_id, done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -251,3 +285,33 @@ def put(self, topic_id, assignee_id):
+class TopicPublishHandler(base.BaseHandler):
+    """A handler for publishing and unpublishing topics."""
+
+    @acl_decorators.can_manage_rights_for_topic
```
can_change_topic_publication_status

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -251,3 +285,33 @@ def put(self, topic_id, assignee_id):
+            raise self.UnauthorizedUserException(e)
+
+        self.values.update({
+            'topic_status_updated': True
```
Is this actually needed? Wouldn't a 200 status code mean that everything worked fine?

aks681 replied:
Yeah, done.

seanlip commented on core/controllers/topic_editor_test.py:
```diff
@@ -348,3 +348,71 @@ def test_assign_topic_manager_role(self):
+
+    def test_publish_and_unpublish_topic(self):
+        """Test the publish and unpublish functionality.
+        """
```
Move to end of previous line. Ditto elsewhere.

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor_test.py:
```diff
@@ -348,3 +348,71 @@ def test_assign_topic_manager_role(self):
+            self.assertFalse(topic_rights.topic_is_published)
+            self.logout()
+
+
```
Drop unnecessary newline. /cc @apb7 @kevinlee12

aks681 replied:
Done.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -779,9 +781,18 @@ def to_dict(self):
+            topic_id: str. The id of the topic.
+            manager_ids: list(str). The id of the users who have been assigned
+                as managers for the topic.
+            topic_is_published: boo. Whether the topic is playable by a learner.
```
bool

Also, I'm not sure that a learner "plays" topics. Perhaps playable --> viewable.

aks681 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
+
+    Raises:
+        Exception. The given topic does not exist.
+        Exception. The user does not have enought rights to publish the topic.
```
enought --> enough

aks681 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
+    user_actions_info = user_services.UserActionsInfo(committer_id)
+    if not check_can_edit_topic(user_actions_info, topic_rights):
+        raise Exception(
+            'The user does not have enought rights to publish the topic.')
```
enought --> enough

aks681 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
+            'The user does not have enought rights to publish the topic.')
+
+    topic_rights.topic_is_published = True
+    commit_cmds = [{
```
We should be using TopicRightsChange objects for these, right? Ditto below.

aks681 replied:
Yeah, created it.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsBackendApiService.js:
```diff
@@ -0,0 +1,145 @@
+    };
+
+    var _isCached = function(topicId) {
+      return topicId in topicRightsCache;
```
`return topicRightsCache.hasOwnProperty(topicId);`

(Btw, for these comments in general, worth changing the source too if this is based on something else in the codebase. Better to make it all consistent.)

aks681 replied:
Done.

seanlip reviewed (changes requested):
Partial review so far, up to TopicRightsBackendApiService.js. Will continue later!

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+      return this._canEditTopic;
+    };
+
+    TopicRights.prototype.getIsPublished = function() {
```
topicRights.isPublished()

If you want to avoid collision with the backend dict / private names, consider using this._published and this._editableByCurrentUser instead in both of those.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,84 @@
+      return this._topicId;
+    };
+
+    TopicRights.prototype.getCanEditTopic = function() {
```
topicRights.canEditTopic() or topicRights.canEdit() or topicRights.isEditable()

(can is already a verb)

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorStateService.js:
```diff
@@ -0,0 +1,198 @@
+    var _topic = TopicObjectFactory.createEmptyTopic();
+    var _topicRights = TopicRightsObjectFactory.createEmptyTopicRights();
+    var _topicIsInitialized = false;
+    var _isLoadingTopic = false;
```
_topicIsLoading or _topicIsCurrentlyLoading

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorStateService.js:
```diff
@@ -0,0 +1,198 @@
+    var _topicRights = TopicRightsObjectFactory.createEmptyTopicRights();
+    var _topicIsInitialized = false;
+    var _isLoadingTopic = false;
+    var _isSavingTopic = false;
```
_topicIsBeingSaved or _topicIsCurrentlyBeingSaved

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditor.js:
```diff
@@ -21,8 +21,14 @@ oppia.constant(
-  function() {
+oppia.controller('TopicEditor', ['$scope', 'TopicEditorStateService',
+  function($scope, TopicEditorStateService) {
+    TopicEditorStateService.loadTopic(GLOBALS.topicId);
```
Try to avoid using GLOBALS if possible. Have a TopicIdService that retrieves the topic ID from the URL. Ditto elsewhere.

/cc @vojtechjelinek @1995YogeshSharma @BenHenning -- can you confirm that the above suggestion is the correct way to handle this?

aks681 replied:
Even in collections, this is done, should I change both places?

seanlip replied:
Yes, I think so. Though would be good to get confirmation from at least one of @vojtechjelinek @1995YogeshSharma @BenHenning first.

vojtechjelinek replied:
Yes, if it's possible, then in both case retrieve the data from the URL.

BenHenning replied:
I actually think we used this with collections as an alternative to parsing the URL like the exploration player does. I found it a bit cleaner since the ID was passed directly from the backend. That being said, I think we also isolated the GLOBALS usage to the controller's constructor and otherwise passed around a collectionId variable where we needed it.

I think I prefer not parsing the URL since it seems a bit safer/more reliable to depend on the exact value passed by the backend rather than parsing out the URL.

seanlip replied:
Ah, @BenHenning -- I think we're working to get rid of GLOBALS across the site. Do you have alternative suggestions for how to handle this if GLOBALS is not available? I'm not sure I see an alternative...

Though, @vojtechjelinek @1995YogeshSharma, this brings up a good point that we should find a way to load the 404 error page if the ID in a URL is invalid and this is discovered on ajax load...

BenHenning replied:
Is the idea to get rid of GLOBALS because the global state being shared here is a bit awkward, or because of the broader Jinja work? If the latter, don't we still need a mechanism for posting data to the page outside of the URL?

I don't have a strong opinion here since the argument makes sense for both approaches.

seanlip replied:
The latter. And to the follow-up question, I think we don't, actually. We'll use subsequent AJAX calls to get the necessary data from the backend.

seanlip replied:
(Though please let us know if you have follow-on thoughts to this!)

seanlip commented on core/templates/dev/head/pages/topic_editor/main_editor/TopicEditorTabDirective.js:
```diff
@@ -0,0 +1,67 @@
+            EVENT_TOPIC_INITIALIZED, EVENT_TOPIC_REINITIALIZED) {
+          $scope.topic = TopicEditorStateService.getTopic();
+          $scope.topicRights = TopicEditorStateService.getTopicRights();
+          $scope.canEditName = GLOBALS.isAdmin;
```
Don't do this, refer to $scope.topicRights and make sure that that has sufficient info to determine this.

aks681 replied:
In the docs, it was written that topic managers also won't be allowed to change the name, so shall I send an isAdmin also in topic_rights?

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/main_editor/TopicEditorTabDirective.js:
```diff
@@ -0,0 +1,67 @@
+          $scope.$on(EVENT_TOPIC_INITIALIZED, _initEditor);
+          $scope.$on(EVENT_TOPIC_REINITIALIZED, _initEditor);
+
+          $scope.switchTopicNameEditor = function() {
```
Split into $scope.openTopicNameEditor and $scope.closeTopicNameEditor

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/topic_editor.html:
```diff
@@ -1,5 +1,49 @@
+
+{% block content %}
+<div ng-controller="TopicEditor">
+  <div>
```
Probably don't need this div?

aks681 replied:
Yeah, done.

seanlip commented on feconf.py:
```diff
@@ -644,6 +644,8 @@ def get_empty_ratings():
 TOPIC_EDITOR_URL_PREFIX = '/topic_editor'
 TOPIC_MANAGER_RIGHTS_URL_PREFIX = '/rightshandler/assign_topic_manager'
+TOPIC_STATUS_URL_PREFIX = '/rightshandler/change_topic_status'
+TOPIC_RIGHTS_URL_PREFIX = '/rightshandler/get_topic_rights'
```
Swap the above two lines with each other to preserve alphabetical order of constant names.

aks681 replied:
Done.

seanlip reviewed (changes requested):
Reviewed the rest. In general, this looks very good!

aks681 committed:
Fixed merge conflicts

aks681 committed:
Added save changes functionality

aks681 commented:
@seanlip PTAL. Added the Save Topic functionality. Also, restricted edit name and publish access to admins. Also, added the Save Changes button to the UI screenshot and the tests for the services.
The topic status is also now visible on the dashboard:
![screenshot from 2018-06-17 00-30-22](https://user-images.githubusercontent.com/22347970/41501641-00134f8a-71c6-11e8-957e-86356ca8b4a8.png)

aks681 committed:
added topic status to dashboard

aks681 committed:
Added tests

aks681 committed:
fixed errors

aks681 committed:
modified topic editor roles

aks681 commented:
Also, modified the permissions so that any topic manager can view any topic editor, but only one assigned to that topic can actually edit it.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -63,7 +64,7 @@ def post(self, topic_id):
     """The editor page for a single topic."""
 
-    @acl_decorators.can_edit_topic
+    @acl_decorators.can_visit_any_topic_editor
```
Perhaps can_view instead of can_visit -- anyone can navigate to the page, but we control whether they can view it or not.

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +228,35 @@ def delete(self, topic_id):
+        topic_rights = topic_services.get_topic_rights(topic_id, strict=False)
+        if topic_rights is None:
+            raise self.InvalidInputException(
+                Exception('Expected a valid topic id to be provided.'))
```
Hm. I thought we usually just pass a string to our custom exceptions, not an Exception object.

aks681 replied:
Yeah, done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -224,6 +228,35 @@ def delete(self, topic_id):
+
+        can_publish_topic = False
+        if role_services.ACTION_PUBLISH_TOPIC in user_actions_info.actions:
+            can_publish_topic = True
```
```
can_publish_topic = (
    role_services.ACTION_PUBLISH_TOPIC in user_actions_info.actions)
```

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -246,8 +279,30 @@ def put(self, topic_id, assignee_id):
+    @acl_decorators.can_change_topic_publication_status
+    def put(self, topic_id):
+        """Publishes or unpublishes a topic.
+        """
```
Move to end of previous line. /cc @apb7 @kevinlee12

aks681 replied:
Done.

seanlip commented on core/controllers/topic_editor_test.py:
```diff
@@ -348,3 +365,64 @@ def test_assign_topic_manager_role(self):
+            self.assertEqual(json_response['can_publish_topic'], True)
+            self.logout()
+
+
```
Drop one of these newlines. /cc @apb7 @kevinlee12

aks681 replied:
Done.

seanlip commented on core/controllers/topics_and_skills_dashboard.py:
```diff
@@ -59,6 +59,13 @@ def get(self):
             summary.to_dict() for summary in skill_summaries]
 
+        topic_rights_list = topic_services.get_all_topic_rights()
+        for topic_rights in topic_rights_list:
```
Seems like O(N^2). Why not make a dict from topic_id to topic_rights object (maybe even have the function return that) and then you can just look up by ID?

aks681 replied:
Yeah, I had thought since number of topics was less, it wouldn't be a problem. Changed it to use dict.

seanlip replied:
Ah -- I think the number of subtopics within a topic is always bounded, but I can see us having a very large number of topics in the future.

seanlip commented on core/domain/acl_decorators.py:
```diff
@@ -1031,6 +1031,30 @@ def test_can_access_topics_and_skills_dashboard(self, **kwargs):
 
+def can_visit_any_topic_editor(handler):
+    """Decorator to check whether the user can access the topics and skills
+    dashboard.
```
Please double-check your docstrings.

aks681 replied:
Done.

seanlip commented on core/domain/acl_decorators.py:
```diff
@@ -1051,3 +1075,25 @@ def test_can_manage_topic_rights(self, topic_id, **kwargs):
+        user_actions_info = user_services.UserActionsInfo(self.user_id)
+
+        if (
+                role_services.ACTION_PUBLISH_TOPIC in
```
"publish topic" seems different from "change topic publication status", since the latter allows unpublishing as well...

aks681 replied:
Ok, will rename to CHANGE_TOPIC_STATUS.

aks681 replied:
Done.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -137,6 +139,63 @@ def to_dict(self):
+                value for 'cmd' is. The possible values for 'cmd' are listed
+                below, together with the other keys in the dict:
+                - 'change_role' (with assignee_id, new_role
+                and old_role)
```
I think you can move this to the previous line.

aks681 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
+    if topic_rights is None:
+        raise Exception('The given topic does not exist')
+    user_actions_info = user_services.UserActionsInfo(committer_id)
+    if not check_can_edit_topic(user_actions_info, topic_rights):
```
"can edit topic" seems different to me than "can publish topic"

aks681 replied:
Yeah, changed it to check for ACTION_CAN_CHANGE_TOPIC_STATUS.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
+    if topic_rights is None:
+        raise Exception('The given topic does not exist')
+    user_actions_info = user_services.UserActionsInfo(committer_id)
+    if not check_can_edit_topic(user_actions_info, topic_rights):
```
Ditto.

aks681 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -694,10 +694,65 @@ def get_topic_rights_from_model(topic_rights_model):
     )
 
 
+def publish_topic(topic_id, committer_id):
```
Should you also check that the topic is not already published? Ditto for unpublished.

aks681 replied:
Ok, done.

seanlip commented on core/domain/topic_services_test.py:
```diff
@@ -549,6 +549,28 @@ def test_admin_can_manage_topic(self):
+            'The user does not have enough rights to publish the topic.'):
+            topic_services.publish_topic(self.TOPIC_ID, self.user_id_a)
+
+        with self.assertRaisesRegexp(
```
Seems possibly iffy, because the topic is unpublished at this point. So even a user with rights could not unpublish it anyway.

aks681 replied:
Done, moved this check to earlier in the test.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+      return this._published;
+    };
+
+    TopicRights.prototype.getCanPublishTopic = function() {
```
canPublishTopic / isTopicPublishable

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+    };
+
+    // Currently only admins can publish/unpublish a topic or edit it's name.
+    TopicRights.prototype.getCanEditName = function() {
```
canEditName / isNameEditable

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+oppia.factory('TopicRightsObjectFactory', [
+  function() {
+    var TopicRights = function(topicRightsObject) {
+      this._published = topicRightsObject.is_published;
```
Why not also call it "published" in the backend?

aks681 replied:
Ok, renamed it.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+      if (this._canPublishTopic) {
+        this._published = true;
+      } else {
+        throw new Error('User is not allowed to edit this topic.');
```
"edit" seems wrong.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+      return this._canPublishTopic;
+    };
+
+    // Sets _isPublished to true only if the user can edit the
```
"edit" seems wrong. Please proofread docstrings.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+      }
+    };
+
+    // Sets _isPublished to false if user can edit the topic.
```
"edit" seems wrong. Please proofread docstrings.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+      if (this._canPublishTopic) {
+        this._published = false;
+      } else {
+        throw new Error('User is not allowed to edit this topic.');
```
"edit" seems wrong.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+
+oppia.factory('TopicRightsObjectFactory', [
+  function() {
+    var TopicRights = function(topicRightsObject) {
```
This is weird -- the constructor should take individual args, not an "object". After all, you're supposed to be _creating_ an object here...

aks681 replied:
Oh, this was supposed to be the topicRightsBackendObject, will rename it.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -0,0 +1,89 @@
+    // Create a new, empty topic rights object. This is not guaranteed to
+    // pass validation tests.
+    TopicRights.createEmptyTopicRights = function() {
+      return new TopicRights({});
```
This doesn't seem right. It makes everything undefined. Can we have a "no rights" default object instead (cannot publish, cannot edit, private) or a null rights object? (Any thoughts on which is better, and why?)

aks681 replied:
I guess a null would be better, as a 'no rights' object is a possible object for a topic manager not assigned to any topic, while null would make a generalized blank object. Will change that.

aks681 replied:
Done.

seanlip replied:
That's good reasoning; thanks!

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorNavbarDirective.js:
```diff
@@ -0,0 +1,99 @@
+            }
+            TopicRightsBackendApiService.publishTopic($scope.topicId).then(
+              function() {
+                $scope.topicRights.publishTopic();
```
Oh, interesting. So it's not actually doing the publishing? In that case, better to do $scope.topicRights.markTopicAsPublished().

aks681 replied:
Ok, done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorNavbarDirective.js:
```diff
@@ -0,0 +1,99 @@
+              });
+          };
+
+          $scope.getChangeListCount = function() {
```
I think this should probably be getChangeListLength, since otherwise it sounds like "number of change lists".

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorNavbarDirective.js:
```diff
@@ -0,0 +1,99 @@
+          };
+
+          $scope.saveChanges = function() {
+            var isPublished = $scope.topicRights.isPublished();
```
topicIsPublished (avoid starting with verb, for non-functions)

aks681 replied:
Ok, done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorNavbarDirective.js:
```diff
@@ -0,0 +1,99 @@
+            }
+            TopicRightsBackendApiService.unpublishTopic($scope.topicId).then(
+              function() {
+                $scope.topicRights.unpublishTopic();
```
markTopicAsUnpublished

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_list_directive.html:
```diff
@@ -38,6 +40,12 @@
       </td>
+      <td>
+        <a ng-href="<[getTopicEditorUrl(topic.id)]>" class="list-summary">
+          <span ng-if="topic.is_published"> Published! </span>
```
Don't need exclamation mark :) It's fun the first time, but could get jarring if you have lots of published topics.

aks681 replied:
Ok, done.

aks681 committed:
made review changes

seanlip commented on core/controllers/topics_and_skills_dashboard.py:
```diff
@@ -59,12 +59,12 @@ def get(self):
+        for topic_summary in topic_summary_dicts:
+            topic_rights = topic_rights_dict[topic_summary['id']]
+            if topic_rights:
+                topic_summary[
```
For the future: you can also do

    topic_summary['is_published'] = (
        topic_rights.topic_is_published)

That said ... shouldn't this be topic_summary['topic_is_published'], similar to above?

aks681 replied:
Yeah, done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -818,13 +824,14 @@ def get_all_topic_rights():
     Returns:
-        list(TopicRights). The list of right objects of all topics present in
-            the datastore.
+        dict. The dict of right objects of all topics present in
```
right objects --> rights objects

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -19,10 +19,10 @@
-      this._published = topicRightsObject.is_published;
-      this._canPublishTopic = topicRightsObject.can_publish_topic;
-      this._canEditTopic = topicRightsObject.can_edit_topic;
+    var TopicRights = function(topicRightsBackendObject) {
```
No, still don't do this -- it's fragile especially if the backend dict changes. Pass each arg separately instead. Let createFromBackendDict parse and interpret the backend object, the constructor shouldn't have to do this.

(Also FYI I think we would call it a backend dict rather than a backend object. I know object is the correct term in JS but that word is pretty generic/overloaded, so that's why we refer to dicts even in the frontend.)

aks681 replied:
Ok, so create would have the backend dict as the argument, and it will split and give the separate fields to the constructor?

seanlip replied:
Yep! (Or, if you're doing that, make sure that "create" is called "createFromBackendDict".)

aks681 replied:
Ok.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -35,31 +35,31 @@ oppia.factory('TopicRightsObjectFactory', [
       return this._canPublishTopic;
     };
 
     // Currently only admins can publish/unpublish a topic or edit it's name.
```
it's --> its

(its is the possessive; it's means "it is" or "it has")

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicRightsObjectFactory.js:
```diff
@@ -75,13 +75,17 @@ oppia.factory('TopicRightsObjectFactory', [
     // pass validation tests.
     TopicRights.createEmptyTopicRights = function() {
-      return new TopicRights({});
+      return new TopicRights({
```
Just to check: why an object with fields that have a null value, rather than saying `topicRights = null`? What are the pros and cons?

Also, for a topic rights object like this, what happens if you call topicRights.markTopicAsUnpublished(), topicRights.canEdit(), etc.? That case needs to be handled explicitly, but then it's going to introduce a lot of clutter. So consider the alternative of having topicRights just not exist at all at the outset; what happens then?

If you do need a topicRights early on then maybe createDefault does need to be a valid rights object but with the most restrictive settings possible.

Anyway -- this is something to figure out now, since you'll need to reuse this pattern in the future.

aks681 replied:
In TopicEditorStateService, if I directly set _topicRights to null, then copyFromTopicRights (which was what was called to set the topic) would throw error. Actually, just createFromBackendDict is required, and the local topicRights object could be set directly from backend dict, instead of doing the copying (which, I think was the reason, that a 'create empty' was created in the first place). Shall I remove the create empty then?

seanlip replied:
Yep, that sounds good to me. Having the actual information from the get-go would be better.

aks681 replied:
Done, added a createInterstitialObject function that returns a valid but highly restrictive object for rights and one with placeholder values for topic object.

seanlip reviewed (changes requested):
Thanks @aks681! Did a pass.

aks681 committed:
made review changes

aks681 committed:
removed GLOBALS

seanlip commented on core/templates/dev/head/domain/topic/TopicObjectFactory.js:
```diff
@@ -285,10 +285,12 @@ oppia.factory('TopicObjectFactory', ['SubtopicObjectFactory',
+    // Create an interstitial topic that would be displayed in the editor until
+    // the actual topic is fetched from the backend.
+    Topic.createInterstitialTopic = function() {
       return new Topic({
```
Sorry I missed this earlier: could you make the topic constructor have individual args, same as for TopicRights?

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topic_editor/TopicEditorNavbarDirective.js:
```diff
@@ -23,17 +23,17 @@ oppia.directive('topicEditorNavbar', [
         '/pages/topic_editor/topic_editor_navbar_directive.html'),
       controller: [
-        '$scope', '$uibModal', 'AlertsService',
+        '$scope', '$uibModal', '$rootScope', 'AlertsService',
```
Why add $rootScope? Don't see it being used...

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/services/contextual/UrlService.js:
```diff
@@ -46,6 +46,10 @@ oppia.factory('UrlService', ['$window', function($window) {
     },
+    getTopicIdFromUrl: function() {
+      var pathname = this.getPathname();
+      return pathname.split('/')[2].substring(0, 12);
```
Add unit tests for this.

Also, why wouldn't it be the whole of the second component? Better to take the second component, check whether it has length 12, throw an error if not ("The URL xxx is not a valid topic URL"), and otherwise return it.

aks681 replied:
I thought of that, but using tabs, I saw hashes in front of the id. So, to ignore that, used the first 12 characters.

aks681 replied:
Done, added tests.

aks681 replied:
I am merging now, will change in this PR (https://github.com/oppia/oppia/pull/5119) if something else can be done.

seanlip replied:
Read up on the structure of URLs -- the hash stuff is a different part, and can be separated out (window.location.hash I think). So, extract only the main part of the URL (without the hash), and take the relevant component of that (validating as described above). This needs to be fixed in a follow-up PR.

seanlip reviewed (approved):
Hi @aks681, just a few more comments. Feel free to address them and then submit. Thanks!

aks681 committed:
made review changes

aks681 committed:
fixed errors
